File: 2018-02/2018-02-22/leetcode121.java
/**
 * date: 2018-02-22
 * author: pwxcoo
 * complexity: T = O(n), S = O(1)
 * describe:    dp
 */

class Solution {
    public int maxProfit(int[] prices) {
        int res = 0;
        for(int i = 1; i < prices.length; i++)
        {
            if(prices[i] > prices[i - 1]) 
            {
                res = Math.max(res, prices[i] - prices[i - 1]);
                prices[i] = prices[i - 1];
            }
        }
        return res;
    }
}

File: 2018-08/2018-08-06/leetcode121.cpp
class Solution
{
  public:
    int maxProfit(vector<int> &prices)
    {
        int s = 0, dif = 0;
        for (int i = s + 1; i < prices.size(); i++)
        {
            if (prices[i] > prices[s])
            {
                if (prices[i] - prices[s] > dif)
                    dif = prices[i] - prices[s];
            }
            else
                s = i;
        }
        return dif;
    }
};